kvdb/src/lib.rs

```diff
@@ -167,9 +181,425 @@ pub trait KeyValueDB: Sync + Send {
+		DatabaseWithCache::flush(self)
+	}
+
+	fn restore(&self, new_db: &str) -> io::Result<()> {
+		DatabaseWithCache::restore(self, new_db)
+	}
+
+	fn iter<'a>(&'a self, col: Option<u32>) -> Box<Iterator<Item=iter::KeyValuePair> + 'a> {
+		let unboxed = DatabaseWithCache::iter(self, col);
+		Box::new(unboxed)
+	}
+
+	fn iter_from_prefix<'a>(&'a self, col: Option<u32>, prefix: &'a [u8])
+		-> Box<Iterator<Item=iter::KeyValuePair> + 'a>
+	{
+		let unboxed = DatabaseWithCache::iter_from_prefix(self, col, prefix);
+		Box::new(unboxed)
+	}
+}
+
+impl<DB> DatabaseWithCache<DB>
+where
+	DB: OpenHandler<DB> + TransactionHandler,
+	for<'a> &'a DB: IterationHandler,
+{
+	/// Helper to create a new transaction.
+	pub fn transaction(&self) -> DBTransaction { DBTransaction::new() }
+	/// Commit transaction to database.
+	pub fn write_buffered(&self, tr: DBTransaction) {
+		let mut overlay = self.overlay.write();
+		let ops = tr.ops;
+		for op in ops {
+			match op {
+				DBOp::Insert { col, key, value } => {
+					let c = Self::to_overlay_column(col);
+					overlay[c].insert(key, KeyState::Insert(value));
+				},
+				DBOp::Delete { col, key } => {
+					let c = Self::to_overlay_column(col);
+					overlay[c].insert(key, KeyState::Delete);
+				},
+			}
+		};
+	}
+
+	/// Commit transaction to database.
+	pub fn write(&self, tr: DBTransaction) -> io::Result<()> {
+		match *self.db.read() {
+			Some(ref db) => {
+				let mut txn = db.write_transaction();
+				let ops = tr.ops;
+				for op in ops {
+					let c = Self::to_overlay_column(op.col());
+					// remove any buffered operation for this key
+					self.overlay.write()[c].remove(op.key());
+
+					match op {
+						DBOp::Insert { key, value, .. } => txn.put(c, &key, &value)?,
+						DBOp::Delete { key, .. } => txn.delete(c, &key)?,
+					};
+				}
+
+				txn.commit()
+			},
+			None => Err(other_io_err("Database is closed")),
+		}
+	}
+
+	/// Get value by key.
+	pub fn get(&self, col: Option<u32>, key: &[u8]) -> io::Result<Option<DBValue>> {
+		match *self.db.read() {
+			Some(ref db) => {
+				let c = Self::to_overlay_column(col);
+				let overlay = &self.overlay.read()[c];
+				match overlay.get(key) {
+					Some(&KeyState::Insert(ref value)) => Ok(Some(value.clone())),
+					Some(&KeyState::Delete) => Ok(None),
+					None => {
+						let flushing = &self.flushing.read()[c];
+						match flushing.get(key) {
+							Some(&KeyState::Insert(ref value)) => Ok(Some(value.clone())),
+							Some(&KeyState::Delete) => Ok(None),
+							None => {
+								let txn = db.read_transaction();
+								txn.get(c, &key)
+							},
+						}
+					},
+				}
+			},
+			None => Ok(None),
+		}
+	}
+
+	/// Get value by partial key. Prefix size should match configured prefix size. Only searches flushed values.
+	pub fn get_by_prefix(&self, col: Option<u32>, prefix: &[u8]) -> Option<Box<[u8]>> {
+		match self.iter_from_prefix(col, prefix).next() {
+			Some((k, v)) => if k.starts_with(prefix) { Some(v) } else { None },
```

Comment:
Why is this `starts_with` check necessary? Can LMDB return values that do not match the prefix?

Reply:
It is actually coming from rocksdb implementation (https://github.com/paritytech/parity-common/blob/1409ddaf07584e0ddaf37c20fccc4a06ae8950fe/kvdb-rocksdb/src/lib.rs#L541).

Reply:
Interesting. I guess the same question applies! Doesn't rocksdb support prefix matching?

If it's a rocksdb quirk, maybe we can add a `TODO` to clean it up?

Reply:
https://github.com/facebook/rocksdb/blob/120bc4715b7e851b32f9b880f5ffac1d4f3a5d50/include/rocksdb/options.h#L1148-L1154

https://github.com/paritytech/parity-common/blob/1409ddaf07584e0ddaf37c20fccc4a06ae8950fe/kvdb-rocksdb/src/lib.rs#L540

Reply:
For posterity, this seems to be fixed in [upstream rocksdb](https://docs.rs/rocksdb/0.12.1/rocksdb/struct.ReadOptions.html), from which we have diverged quite a bit I believe. Oddly enough the [parity-rocksdb fork](https://github.com/paritytech/rust-rocksdb/blob/master/README.md) names prefix search as one of the reasons for forking, and yet I can't find it in our code.